File: pbmx_ffi/java/io/rmf/pbmx/payloads/OpenStackPayload.java
package io.rmf.pbmx.payloads;

import io.rmf.pbmx.BlockBuilder;
import io.rmf.pbmx.Id;
import io.rmf.pbmx.Mask;
import java.util.Collection;
import java.util.ArrayList;

public final class OpenStackPayload extends Payload {
    public OpenStackPayload(Collection<Mask> stack) {
        this.stack = new ArrayList(stack);
    }

    @Override
    public void addTo(BlockBuilder builder) {
        builder.openStack(stack);
    }

    private final Collection<Mask> stack;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/payloads/OpenStackPayload.kt
package io.rmf.pbmx.payloads

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import io.rmf.pbmx.Mask
import io.rmf.pbmx.ffi.FFI
import io.rmf.pbmx.ffi.RawMask
import io.rmf.pbmx.jnaArrayOf

class OpenStackPayload(handle: Pointer) : Payload(handle) {
    val stack: Collection<Mask>

    init {
        val length = LongByReference()
        var r = FFI.pbmx_unwrap_open_stack(this.handle, null, length)
        assert(r == 0)

        val outMasks = jnaArrayOf(RawMask(), length.value.toInt())
        r = FFI.pbmx_unwrap_open_stack(this.handle, outMasks, length)
        assert(r != 0)
        this.stack = outMasks.map { Mask(it) }.toList()
    }
}
